extension/src/frequency.rs

```diff
@@ -682,6 +1049,98 @@ fn topn_internal(
     })
 }
 
+#[pg_extern(immutable, parallel_safe, name="topn", schema = "toolkit_experimental")]
+pub fn topn_bigint(
+    agg: SpaceSavingBigIntAggregate<'_>,
+    n: i32,
+) -> impl std::iter::Iterator<Item = i64> + '_ {
+    validate_topn_for_topn_agg(n, agg.topn, agg.freq_param, agg.values_seen, agg.counts.iter());
+    let min_freq = if agg.topn == 0 {
+        agg.freq_param
+    } else {
+        0.
+    };
+
+    topn_bigint_internal(agg, n, min_freq)
+}
+
+#[pg_extern(immutable, parallel_safe, name="topn", schema = "toolkit_experimental")]
+pub fn default_topn_bigint(
+    agg: SpaceSavingBigIntAggregate<'_>,
+) -> impl std::iter::Iterator<Item = i64> + '_ {
+    if agg.topn == 0 {
+        pgx::error!("frequency aggregates require a N parameter to topn")
+    }
+    let n = agg.topn as i32;
+    topn_bigint(agg, n)
+}
+
+fn topn_bigint_internal(
+    agg: SpaceSavingBigIntAggregate<'_>,
+    max_n: i32,
+    min_freq: f64,
+) -> impl std::iter::Iterator<Item = i64> + '_ {
+    let iter = agg
```

Comment:
This is pretty complex, what with zipping two things together and having 3 different exit criteria. It would be great not to duplicate that algorithm. I show how to factor it out in commit d6e293a9ae9d6a5ae7f1f880feb792d8b5b829a4.

Reply:
Yeah, even cleaner with the new change.